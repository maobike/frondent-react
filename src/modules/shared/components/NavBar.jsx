import React, { useState } from 'react';

import { NavLink } from 'react-router-dom';

import {
    Collapse,
    Navbar,
    NavbarToggler,
    NavbarBrand,
    Nav,
} from 'reactstrap';


function NavBar() {
    const [isOpen, setIsOpen] = useState(false);

    const toggle = () => setIsOpen(!isOpen);

    return (
        <>
            <Navbar color="dark" dark={true} expand="md">
                <NavbarBrand className='text-warning' data-testid='titleApp'>Test Corserva</NavbarBrand>
                <NavbarToggler onClick={toggle} />
                <Collapse isOpen={isOpen} navbar>
                    <Nav className="me-auto" navbar>
                        <NavLink
                            to="/"
                            className={({ isActive }) => `nav-item nav-link  ${isActive ? 'active' : ''}`}
                        >
                            Usuarios
                        </NavLink>
                        <NavLink
                            to="other"
                            className={({ isActive }) => `nav-item nav-link  ${isActive ? 'active' : ''}`}
                        >
                            Art√≠culos
                        </NavLink>
                    </Nav>
                </Collapse>
            </Navbar>
        </>
    );
}

export default NavBar;