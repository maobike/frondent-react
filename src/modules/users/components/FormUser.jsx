import React, { Fragment, useState, useEffect } from 'react'
import { Row, Col, FormGroup, Label, Button, ModalHeader, ModalBody, ModalFooter } from "reactstrap";
import { useForm } from 'react-hook-form';
import '../../../assets/font-awesome';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import UserService from '../services/UserService'
import CategoryService from '../../../modules/categories/services/CategoryService'
import Swal from 'sweetalert2';

export function FormUser(props) {
    const { register, handleSubmit, setValue, setError, formState: { errors } } = useForm();
    const { toggle, user, refreshList } = props;
    const [title, setTitle] = useState(null);
    const [categories, setCategories] = useState(null);

    // Función para saber si viene a crear o editar el cliente
    useEffect(() => {
        const getTitle = () => {
            if (user) {
                setTitle('Editar Usuario');
                try {
                    setValue("name", user.name)
                    setValue("email", user.email)
                    setValue("phone", user.phone)
                    setValue("status", user.status)
                } catch (error) {
                    console.log(error);
                }
            } else {
                setTitle('Crear Usuario');
            }
        };

        getTitle();
    }, []);

    // Función para enviar a guardar
    const onSubmit = async (data) => {
        try {
            // Crear o editar
            if (user) {
                updateUser(data);
            } else {
                addUser(data);
            }
        } catch (error) {
            console.log(error);
        }
    }

    // Crear cliente
    const addUser = async (data) => {
        const result = await UserService.addUser(data);
        try {
            if (result.data?.userId) {
                setError("dni", "match", "La cédula del usuario ya existe.");
                return false;
            } else {
                if (result.status === 200) {
                    Swal.fire({
                        title: "Usuario creado con éxito",
                        text: result.data.message,
                        icon: "success",
                        confirmButtonText: "Aceptar",
                    });
                } else {
                    Swal.fire({
                        title: "Transacción errónea",
                        text: result.data,
                        icon: "error",
                        confirmButtonText: "Aceptar",
                    });
                }
                toggle();
            }
        } catch (error) {
            console.log(error);
        } finally {
            refreshList();
        }
    }

    // Actualizar cliente
    const updateUser = async (data) => {
        try {
            const result = await UserService.editUser(user.id, data);
            if (result.status === 200) {
                Swal.fire({
                    title: "Usuario editado con éxito",
                    text: result.data.message,
                    icon: "success",
                    confirmButtonText: "Aceptar",
                });
            } else {
                Swal.fire({
                    title: "Transacción errónea",
                    text: result.data.message,
                    icon: "error",
                    confirmButtonText: "Aceptar",
                });
            }
        } catch (error) {
            console.log(error);
            Swal.fire({
                title: "Transacción errónea",
                text: "Ocurrió un error, por favor intente de nuevo",
                icon: "error",
                confirmButtonText: "Aceptar",
            });
        } finally {
            refreshList();
            toggle();
        }
    }

    return (
        <Fragment>
            <ModalHeader toggle={toggle}>{title}</ModalHeader>
            <ModalBody>
                <form name="form-list" id="form-list">
                    <Row>
                        <Col md={12}>
                            <FormGroup>
                                <Label>Nombre <span className="input-obligatorio">*</span></Label>
                                <input className="form-control"
                                    type="text"
                                    autoFocus
                                    maxLength="100"
                                    name="name"
                                    {...register("name", { 
                                        required: true,
                                        minLength: 5,
                                        maxLength: 100,
                                        pattern: /^[a-zA-ZñÑ\s]*$/i 

                                    })}
                                />
                                <div className="text-error">
                                    {errors.name?.type === "required" && "El nombre es requerido"}
                                    {errors.name?.type === "minLength" && "Este campo debe contener mínimo 5 caracteres"}
                                    {errors.name?.type === "maxLength" && "Este campo debe contener máximo 100 caracteres"}
                                    {errors.name?.type === "pattern" && "Este campo no debe contener números ni caracteres especiales" }

                                </div>
                            </FormGroup>
                        </Col>
                    </Row>
                    <Row>
                        <Col md={12}>
                            <FormGroup>
                                <Label>Email</Label>
                                <input className="form-control"
                                    type="email"
                                    maxLength="150"
                                    name="email"
                                    { ...register("email", { 
                                        required: true,
                                        pattern: /^[\w.%+-]+@[\w.-]+\.[a-zA-Z]{2,}$/i
                                    })}
                                />
                                <div className="text-error">
                                    {errors.email?.type === "required" && "El email es requerido"}
                                    {errors.email?.type === "maxLength" && "Este campo debe contener máximo 150 caracteres"}
                                    {errors.email?.type === "pattern" && "Este campo debe ser un correo valido" }
                                </div>
                            </FormGroup>
                        </Col>
                    </Row>

                    <Row>
                        <Col md={6}>
                            <FormGroup>
                                <Label>Teléfono <span className="input-obligatorio">*</span></Label>
                                <input className="form-control"
                                    type="number"
                                    maxLength="10"
                                    name="phone"
                                    { ...register("phone", { 
                                        required: true,
                                        maxLength: 10,
                                        pattern: /^[0-9]+$/
                                    })}
                                />
                                <div className="text-error">
                                    {errors.phone?.type === "required" && "El teléfono es requerido"}
                                    {errors.phone?.type === "maxLength" && "Este campo debe contener máximo 10 caracteres"}
                                    {errors.phone?.type === "pattern" && "Este campo debe ser un teléfono valido" }
                                </div>
                            </FormGroup>
                        </Col>
                        <Col md={6}>
                            <FormGroup>
                                <Label>estado <span className="input-obligatorio">*</span></Label>
                                <select className="form-control"
                                    name="status"
                                    {...register("status", {
                                        required: true,
                                    })}
                                >
                                    <option value="true">Activo</option>
                                    <option value="false">Inactivo</option>
                                </select>
                                <div className="text-error">
                                    {errors.category_id?.type === "required" && "La categoría es requerida"}
                                </div>
                            </FormGroup>
                        </Col>
                    </Row>

                </form>
            </ModalBody>
            <ModalFooter>
                <Button color="primary" onClick={handleSubmit(onSubmit)}>
                    <FontAwesomeIcon icon="save" /> Guardar
                </Button>
                <Button color="secondary" onClick={toggle}>
                    <FontAwesomeIcon icon="eraser" /> Cancel
                </Button>
            </ModalFooter>
        </Fragment>
    )
}