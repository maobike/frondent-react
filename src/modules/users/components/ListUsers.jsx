import React, { useState, useEffect } from 'react';
import '../../../assets/font-awesome';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button, Card, Row, Col, Table, CardBody, InputGroup, Input, InputGroupText, Modal } from 'reactstrap';

import UserService from './../services/UserService';
import { FormUser } from './FormUser';

export const ListUsers = () => {
    const [dataUser, setDataUser] = useState(null);
    const [dataUserFilter, setDataUserFilter] = useState(null);
    const [modal, setModal] = useState(false);
    const [user, setUser] = useState(null);
    const [refresh, setRefresh] = useState();
    const [inputText, setInputText] = useState('');

    // Búsqueda con paginación a BD
    const SearchUser = async (e) => {
        e.preventDefault();
    };

    useEffect(() => {
        const getUsers = async () => {
            try {
                // Llamamos el servicio
                const result = await UserService.getUsers();
                // Asignamos el valor del resultado
                setDataUser(result.data.users);
                setDataUserFilter(result.data.users);
            } catch (error) {
                console.log(error);
            }
        };
        getUsers();
    }, [refresh]);

    // Búsqueda por correo, nombre, apellido o país
    const handleInputChange = (input) => {
        // limpiar arreglo
        setDataUserFilter(null);
        var newArray = [];
        const { value } = input.target;
        //Cambiar entrada a mayúsculas
        setInputText(value.toUpperCase());

        if (dataUser !== null || dataUser !== '' || dataUser !== undefined) {
            dataUser.map((dataUser, index) => {
                if (
                    dataUser?.name.toUpperCase().includes(value.toUpperCase()) ||
                    dataUser?.email.toUpperCase().includes(value.toUpperCase()) ||
                    dataUser?.phone.toUpperCase().includes(value.toUpperCase())
                ) {
                    newArray.push(dataUser);
                    setDataUserFilter(newArray);
                }
                return dataUser;
            });
        }
    };

    const toggle = () => {
        setModal(!modal);
    };

    const refreshList = () => {
        setRefresh(!refresh);
    };

    // Abrir modal crear o editar
    const formUser = (user) => {
        setUser(user);
        toggle('modal');
    };
    return (
        <div className="content">
            <Row>
                <Col md="12">
                    <Card>
                        <CardBody>
                            <Col md="12" className="">
                                <Table className="">
                                    <tbody>
                                        <tr>
                                            <td>
                                                <form onSubmit={SearchUser}>
                                                    <Row >
                                                        <Col md={4}>
                                                            <InputGroup className="no-border">
                                                                <Input
                                                                    placeholder="Buscar..."
                                                                    autoComplete="off"
                                                                    name="SearchUser"
                                                                    value={inputText}
                                                                    onChange={handleInputChange}
                                                                />
                                                                <InputGroupText>
                                                                    <FontAwesomeIcon icon="search" />
                                                                </InputGroupText>
                                                                <Button color='primary'  onClick={() => formUser()}>
                                                                    <FontAwesomeIcon icon="plus" /> Nuevo Usuario
                                                                </Button>
                                                            </InputGroup>
                                                        </Col>
                                                    </Row>
                                                </form>
                                            </td>
                                        </tr>
                                    </tbody>
                                </Table>
                                <div style={{ overflow: 'auto' }}>
                                    <Table hover striped bordered className="borde-tabla">
                                        <thead>
                                            <tr>
                                                <th>Id</th>
                                                <th>Nombre</th>
                                                <th>Email</th>
                                                <th>Teléfono</th>
                                                <th>Estatus</th>
                                            </tr>
                                        </thead>
                                        <tbody>
                                            {dataUserFilter && dataUserFilter.length > 0 && dataUserFilter.map((dataRes, index) => {
                                                return (
                                                    <tr key={index}>
                                                        <td>{dataRes.id}</td>
                                                        <td>{dataRes.name}</td>
                                                        <td>{dataRes.email}</td>
                                                        <td>{dataRes.phone}</td>
                                                        <td>{(dataRes.status) ? 'Activo' : 'Inactivo'}</td>
                                                        <td style={{ textAlign: 'center', alignSelf: 'stretch' }}>
                                                            <Button title="Editar usuario" color="link" onClick={() => formUser(dataRes)}>
                                                                <FontAwesomeIcon icon="edit" />
                                                            </Button>
                                                        </td>
                                                    </tr>
                                                );
                                            })}
                                        </tbody>
                                    </Table>
                                </div>
                            </Col>
                        </CardBody>
                    </Card>
                </Col>
            </Row>
            <div>
                <Modal isOpen={modal} toggle={toggle} size="xl">
                    <FormUser toggle={toggle} user={user} refreshList={refreshList}></FormUser>
                </Modal>
            </div>
        </div>
    )
}