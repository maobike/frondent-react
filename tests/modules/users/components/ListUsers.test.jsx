import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { ListUsers } from '../../../../src/modules/users/components/ListUsers';
import { BrowserRouter } from 'react-router-dom';

describe('ListUsers Component', () => {
  it('This should render correctly', () => {

    const { container } = render(
      <BrowserRouter>
        <ListUsers />
      </BrowserRouter>
    );

    expect( container).toMatchSnapshot();
  });

//   it('debe manejar cambios en el input de búsqueda', () => {
//     const { getByPlaceholderText } = render(<ListUsers />);
//     const searchInput = getByPlaceholderText('Buscar...');
    
//     // Simula un cambio en el input de búsqueda
//     fireEvent.change(searchInput, { target: { value: 'John' } });
    
//     // Asegúrate de que el valor del input haya cambiado
//     expect(searchInput).toHaveValue('John');
//   });

});
