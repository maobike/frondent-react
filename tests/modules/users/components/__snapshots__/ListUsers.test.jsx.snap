// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListUsers Component This should render correctly 1`] = `
<div>
  <div
    class="content"
  >
    <div
      class="row"
    >
      <div
        class="col-md-12"
      >
        <div
          class="card"
        >
          <div
            class="card-body"
          >
            <div
              class="col-md-12"
            >
              <table
                class="table"
              >
                <tbody>
                  <tr>
                    <td>
                      <form>
                        <div
                          class="row"
                        >
                          <div
                            class="col-md-4"
                          >
                            <div
                              class="no-border input-group"
                            >
                              <input
                                autocomplete="off"
                                class="form-control"
                                name="SearchUser"
                                placeholder="Buscar..."
                                type="text"
                                value=""
                              />
                              <span
                                class="input-group-text"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="svg-inline--fa fa-magnifying-glass "
                                  data-icon="magnifying-glass"
                                  data-prefix="fas"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 512 512"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352a144 144 0 1 0 0-288 144 144 0 1 0 0 288z"
                                    fill="currentColor"
                                  />
                                </svg>
                              </span>
                              <button
                                class="btn btn-primary"
                                type="button"
                              >
                                <svg
                                  aria-hidden="true"
                                  class="svg-inline--fa fa-plus "
                                  data-icon="plus"
                                  data-prefix="fas"
                                  focusable="false"
                                  role="img"
                                  viewBox="0 0 448 512"
                                  xmlns="http://www.w3.org/2000/svg"
                                >
                                  <path
                                    d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
                                    fill="currentColor"
                                  />
                                </svg>
                                 Nuevo Usuario
                              </button>
                            </div>
                          </div>
                        </div>
                      </form>
                    </td>
                  </tr>
                </tbody>
              </table>
              <div
                style="overflow: auto;"
              >
                <table
                  class="borde-tabla table table-bordered table-striped table-hover"
                >
                  <thead>
                    <tr>
                      <th>
                        Id
                      </th>
                      <th>
                        Nombre
                      </th>
                      <th>
                        Email
                      </th>
                      <th>
                        Tel√©fono
                      </th>
                      <th>
                        Estatus
                      </th>
                    </tr>
                  </thead>
                  <tbody />
                </table>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div />
  </div>
</div>
`;
